webpush/encrypt.go

```diff
@@ -139,36 +142,64 @@ type EncryptionResult struct {
 	ServerPublicKey []byte
 }
 
+// Decrypt an encrypted messages.
+func Decrypt(sub *Subscription, crypt *EncryptionResult, subPrivate []byte) (plain []byte, err error) {
+	secret := sharedSecret(curve, crypt.ServerPublicKey, subPrivate)
+	prk := hkdf(sub.Auth, secret, authInfo, 32)
+
+	// Derive the Content Encryption Key and nonce
+	ctx := newContext(sub.Key, crypt.ServerPublicKey)
+	cek := newCEK(ctx, crypt.Salt, prk)
+	nonce := newNonce(ctx, crypt.Salt, prk)
+
+	plain, err = decrypt(crypt.Ciphertext, cek, nonce)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+// TODO: input should be a []byte ( proto, etc )
+
 // Encrypt a message such that it can be sent using the Web Push protocol.
 // You can find out more about the various pieces:
 //    - https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding
 //    - https://en.wikipedia.org/wiki/Elliptic_curve_Diffie%E2%80%93Hellman
 //    - https://tools.ietf.org/html/draft-ietf-webpush-encryption
 func Encrypt(sub *Subscription, message string) (*EncryptionResult, error) {
 	plaintext := []byte(message)
+
+	// Use ECDH to derive a shared secret between us and the client. We generate
+	// a fresh private/public key pair at random every time we encrypt.
+	serverPrivateKey, serverPublicKey, err := randomKey()
+	if err != nil {
+		return nil, err
+	}
+
+	return EncryptWithTempKey(sub, plaintext, serverPrivateKey, serverPublicKey)
+}
+
+// Encrypt a message using Web Push protocol, reusing the temp key.
+// A new salt will be used. This is ~20% faster.
+func EncryptWithTempKey(sub *Subscription, plaintext []byte,
```

Comment:
Changing the signature of an exported function is bad in Go, since it will break existing consumers, so not `Encrypt`. I definitely think that `EncryptWithKey` is the name that makes sense - Ephemeral is the opposite of the correct meaning, since the whole point is to allow you to *not* use a key that disappears as soon as you use it.